docs/design/PicoD-Design.md

```diff
@@ -0,0 +1,458 @@
+# PicoD Design Document
+
+Author: VanderChen
+
+## Motivation
+
+The current AgentCube sandbox implementation relies on SSH (via `ssh_client.py`) for remote code execution, file transfer, and sandbox management. While SSH provides robust authentication and encryption, it introduces several challenges:
+
+1. **Heavyweight Protocol**: SSH requires complex handshake procedures, key management, and session overhead
+2. **Limited Customization**: SSH protocol constraints make it difficult to implement custom authentication schemes or optimize for specific use cases
+
+To address these limitations, we propose **PicoD** (Pico Daemon) - a lightweight, HTTP-based service daemon that provides essential sandbox capabilities with minimal overhead while maintaining security through token-based authentication.
+
+### Design Goals
+
+PicoD is designed as a **stateless daemon** that processes individual gRPC requests independently:
+
+- **Lightweight**: Minimal resource footprint suitable for containerized sandbox environments
+- **Simple Protocol**: Pure gRPC/protobuf-based APIs that are easy to integrate and debug
+- **Secure**: Token-based authentication without preset user requirements
+- **No Lifecycle Management**: PicoD does not manage sandbox lifecycle (creation, deletion, monitoring). These responsibilities belong to the AgentCube control plane.
+- **Single Request Processing**: Each gRPC call (Execute, ReadFile, WriteFile) is handled independently without maintaining state between requests.
+- **No Session Management**: No persistent connections or session state. Each request is authenticated via token in metadata.
+- **Ephemeral Process**: PicoD runs for the lifetime of the sandbox container but does not track or manage the sandbox's lifecycle events.
+
+## Use Case
+
+PicoD enables AI agents to interact with sandboxed environments through the AgentCube SDK. The following example demonstrates a complete workflow using multiple PicoD APIs:
+
+### Machine Learning Workflow
+
+An AI agent performs a complete machine learning workflow - uploading data, installing dependencies, training a model, and downloading results:
+
+```python
+from agentcube import Sandbox
+
+# Create a sandbox instance
+sandbox = Sandbox(ttl=3600, image="python:3.11-slim")
```

Comment:
This pr https://github.com/volcano-sh/agentcube/pull/19/files is abstracting a codeInterpreterClient class